src/helpers/tx/estimateTx/estimateTx.ts

```diff
@@ -0,0 +1,25 @@
+import { ConnectedWallet } from "@terra-money/wallet-provider";
+import { Estimate, TxContent } from "types/tx";
+import { WalletState } from "contexts/WalletContext";
+import estimateCosmosFee from "./estimateCosmosFee";
+import { estimateEVMFee } from "./estimateEVMfee";
+import estimateTerraFee from "./estimateTerraFee";
+
+export default async function estimateTx(
+  content: TxContent,
+  wallet: WalletState,
+  terraWallet?: ConnectedWallet
+): Promise<Estimate | null> {
+  try {
+    switch (content.type) {
+      case "cosmos":
+        return estimateCosmosFee(wallet, content.val);
+      case "terra":
+        return estimateTerraFee(wallet, terraWallet!, content.val);
+      default:
+        return estimateEVMFee(wallet, content.val);
+    }
+  } catch (err) {
+    return null;
```

Comment:
if you look at the consumers of this function. The `null` case is handled at the point where it is called giving more refined flows

```ts
const estimate = esimateTx()
if(!estimate) return "Simul Failed" 
//done at this point


```



you are proposing 
```ts
return Promise<Estimate>
```
and in the consumers 
```ts
try { 
const estimated = estimateTx()
const tx = send(estimated)

} catch (err} {
  
  // determine what is `err` and show appropriate error

}

```

to make this work we need additional type of `error` say `SimulateError`

and in catch block we do
```ts
// sendError, typeError, other type of errors exist here
if(err instance of SimulateError){
  // Simul failed
} 
```

Reply:
> to make this work we need additional type of error say SimulateError

Might be a good idea. We already have this type of error logic throughout the webapp, see [errors.ts](https://github.com/AngelProtocolFinance/angelprotocol-web-app/blob/ast-polygon/src/errors/errors.ts). The main benefit is less intermediate checks for errors, to use your example it would mean instead of:
```ts
try { 
  const estimated = estimateTx()
  if (!estimated) {
    return // handle error somehow;
  }
  
  const tx = send(estimated) 
  if (!tx || !!tx.error) {
    return // handle error somehow;
  }
  ...
} catch (err} {
  // still have to check and handle any unanticipated errors
}
```

we would have:
```ts
try { 
  const estimated = estimateTx()
  // assume everything is fine, as otherwise `catch` clause comes into play
  const tx = send(estimated)
  // assume everything is fine, as otherwise `catch` clause comes into play
  ...
} catch (err} {
  // determine what is `err` and show appropriate error
  if(err instance of SimulateError){
    // Simul failed
  } 
  if(err instance of SimulateError2){
    // Simul failed
  } 
  ...
}
```
But we already have `ErrorContext.handleError` which does the `if(err instance of SimulateError)` checks. We could modify it or create a new `ErrorContext.handleTxError` function that uses `TxPrompt` to display errors, so we would get something even shorter:
```ts
const { handleTxError } = useErrorContext();
...
try { 
  const estimated = estimateTx()
  // assume everything is fine, as otherwise `catch` clause comes into play
  const tx = send(estimated)
  // assume everything is fine, as otherwise `catch` clause comes into play
  ...
} catch (err} {
  handleTxError(err);
}
```

Reply:
one of the considerations why errors are handled upfront is that not every flow on our app falls into this 
general show `prompt` on error category

here are examples
<img width="536" alt="image" src="https://user-images.githubusercontent.com/89639563/227088890-a6eca37c-ad8c-4326-83cd-9b4e4e2009f5.png">

<img width="1246" alt="image" src="https://user-images.githubusercontent.com/89639563/227088975-97377bd0-d417-42b9-8dc5-babc7dcecf61.png">

Reply:
About using `handleTxError` as part of `ErrorContext`, it was just an example of a wrapper function for `showModal(TxPrompt, ...)` which is used in different files, but it could've been understood even as a local helper function.

Point was that just throwing errors and handling them in one place might result in code that's even easier to read/understand then code with many `if/else`s throughout it.

IMO, this:
```ts
try { 
  const estimated = estimateTx()
  // assume everything is fine, as otherwise `catch` clause comes into play
  const tx = send(estimated)
  // assume everything is fine, as otherwise `catch` clause comes into play
  ...
} catch (err} {
  handleTxError(err);
}
...
function handleTxError(error: any) {
 // handle error
}
```
is a bit clearer than this:
```ts
try { 
  const estimated = estimateTx()
  if (!estimated) {
    return // handle error somehow;
  }
  
  const tx = send(estimated) 
  if (!tx || !!tx.error) {
    return // handle error somehow;
  }
  ...
} catch (err} {
  // still have to check and handle any unanticipated errors
}
```

Maybe this is a matter of personal taste.

Anyway, this was just a suggestion, let's not delve into that any further.